File: library/src/main/java/com/speshiou/android/common/ui/ads/LoadFbNativeAdTask.java
package com.speshiou.android.common.ui.ads;

import android.content.Context;
import android.text.TextUtils;
import android.view.View;
import android.view.ViewGroup;

import com.facebook.ads.Ad;
import com.facebook.ads.AdChoicesView;
import com.facebook.ads.AdError;
import com.facebook.ads.AdIconView;
import com.facebook.ads.AdListener;
import com.facebook.ads.MediaView;
import com.facebook.ads.NativeAd;
import com.facebook.ads.NativeAdListener;
import com.speshiou.android.common.R;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by joey on 2017/9/12.
 */

public class LoadFbNativeAdTask extends LoadAdTask {

    private NativeAd mNativeAd;

    public LoadFbNativeAdTask(Context context, AdViewRecycler adViewRecycler, String unitId) {
        super(context, adViewRecycler, unitId);
    }

    @Override
    protected void onLoad() {
        super.onLoad();
        NativeAd nativeAd = new NativeAd(mContext, mUnitId);
        nativeAd.setAdListener(new NativeAdListener() {

            @Override
            public void onMediaDownloaded(Ad ad) {

            }

            @Override
            public void onError(Ad ad, AdError error) {
//                    Log.e("joey", "[fb] onError error=" + error.getErrorMessage());
                if (mNativeAd != null && mNativeAd.isAdLoaded()) {
                    onLoaded();
                } else {
                    onFailedToLoad();
                }
            }

            @Override
            public void onAdLoaded(Ad ad) {
                if (mNativeAd != null) {
                    mNativeAd.destroy();
                }
                mNativeAd = (NativeAd) ad;
                if (mNativeAd.isAdLoaded()) {
                    mNativeAd.unregisterView();
                }

                onLoaded();
            }

            @Override
            public void onAdClicked(Ad ad) {
                // Ad clicked callback
            }

            @Override
            public void onLoggingImpression(Ad ad) {

            }
        });

        // Request an ad
        nativeAd.loadAd();
    }

    @Override
    protected void detachAdView() {
        super.detachAdView();
        if (mNativeAd == null || !mNativeAd.isAdLoaded()) {
            return;
        }

        mNativeAd.unregisterView();
    }

    @Override
    public void attachAdView(ViewGroup adContainer) {
        super.attachAdView(adContainer);

        if (mNativeAd == null || !mNativeAd.isAdLoaded()) {
            return;
        }

        mNativeAd.unregisterView();

        View adView = mAdViewRecycler.obtainAdView(mContext, AdViewType.AD_FB_NATIVE, mUnitId);
        AdViewHolder viewHolder = (AdViewHolder) adView.getTag();

        viewHolder.viewAdTag.setVisibility(View.VISIBLE);
        AdIconView nativeAdIcon = adView.findViewById(R.id.native_ad_icon);
        MediaView nativeAdMedia = adView.findViewById(R.id.native_ad_media);
        // Set the Text.
        viewHolder.viewAdTag.setText(mNativeAd.getSponsoredTranslation());
        viewHolder.textViewTitle.setText(mNativeAd.getAdvertiserName());
        viewHolder.textViewSubtitle.setVisibility(TextUtils.isEmpty(mNativeAd.getAdSocialContext()) ? View.GONE : View.VISIBLE);
        viewHolder.textViewSubtitle.setText(mNativeAd.getAdSocialContext());
        viewHolder.textViewBody.setVisibility(TextUtils.isEmpty(mNativeAd.getAdBodyText()) ? View.GONE : View.VISIBLE);
        viewHolder.textViewBody.setText(mNativeAd.getAdBodyText());
        viewHolder.buttonAction.setVisibility(mNativeAd.hasCallToAction() ? View.VISIBLE : View.GONE);
        viewHolder.buttonAction.setText(mNativeAd.getAdCallToAction());

        // Add the AdChoices icon
        AdChoicesView adChoicesView = new AdChoicesView(mContext, mNativeAd, true);
        viewHolder.adChoicePlaceHolder.removeAllViews();
        viewHolder.adChoicePlaceHolder.addView(adChoicesView);

        // Register the Title and CTA button to listen for clicks.
        List<View> clickableViews = new ArrayList<>();
        clickableViews.add(adContainer);
        clickableViews.add(viewHolder.buttonAction);
        mNativeAd.registerViewForInteraction(adContainer, nativeAdMedia, nativeAdIcon, clickableViews);

        if (adView.getParent() != null && adView.getParent() instanceof ViewGroup) {
            ((ViewGroup) adView.getParent()).removeView(adView);
        }
        adContainer.addView(adView);
    }
}

File: library/src/main/java/com/speshiou/android/common/ui/ads/LoadFbNativeAdTask.kt
package com.speshiou.android.common.ui.ads

import android.content.Context
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.view.ViewGroup
import com.facebook.ads.*
import com.speshiou.android.common.R
import com.speshiou.android.common.ui.BaseApplication
import java.util.ArrayList

class LoadFbNativeAdTask(context: Context, adViewRecycler: AdViewRecycler, val fbAdType: String, unitId: String) : LoadAdTask(context, adViewRecycler, fbAdType, unitId) {

    private var mNativeAd: NativeAdBase? = null

    override fun onLoad() {
        super.onLoad()
        val nativeAd: NativeAdBase = when (this.adType) {
            AdType.AD_FB_NATIVE -> NativeAd(context, unitId)
            else -> NativeBannerAd(context, unitId)
        }
        nativeAd.setAdListener(object : NativeAdListener {

            override fun onMediaDownloaded(ad: Ad) {

            }

            override fun onError(ad: Ad, error: AdError) {
                //                    Log.e("joey", "[fb] onError error=" + error.getErrorMessage());
                if (mNativeAd != null && mNativeAd?.isAdLoaded == true) {
                    onLoaded()
                } else {
                    onFailedToLoad()
                }

                val bundle = Bundle()
                bundle.putString("ad_type", fbAdType)
                bundle.putString("ad_id", unitId)
                bundle.putString("ad_error_code", error.errorCode.toString())
                BaseApplication.firebaseAnalytics?.logEvent("ad_error", bundle)
            }

            override fun onAdLoaded(ad: Ad) {
                mNativeAd?.destroy()
                mNativeAd = ad as NativeAdBase
                if (ad.isAdLoaded) {
                    ad.unregisterView()
                }

                onLoaded()
            }

            override fun onAdClicked(ad: Ad) {
                AdCompat.logClickEvent(adType, unitId)
            }

            override fun onLoggingImpression(ad: Ad) {
                AdCompat.logImpressionEvent(adType, unitId)
            }
        })

        // Request an ad
        nativeAd.loadAd()
    }

    override fun detachAdView() {
        super.detachAdView()
        if (mNativeAd == null || mNativeAd?.isAdLoaded == false) {
            return
        }

        mNativeAd?.unregisterView()
    }

    public override fun attachAdView(adContainer: ViewGroup) {
        super.attachAdView(adContainer)

        if (mNativeAd == null || mNativeAd?.isAdLoaded == false) {
            return
        }

        val ad = mNativeAd ?: return
        val adViewType = when (adType) {
            AdType.AD_FB_NATIVE -> {
                AdViewType.AD_FB_NATIVE
            } else -> {
                AdViewType.AD_FB_NATIVE_BANNER
            }
        }
        val adView = adViewRecycler.obtainAdView(context, adViewType) ?: return
        ad.unregisterView()


        val viewHolder = adView.tag as AdViewHolder

        viewHolder.viewAdTag?.visibility = View.VISIBLE
        val nativeAdIcon = adView.findViewById<AdIconView>(R.id.native_ad_icon)
        val nativeAdMedia = adView.findViewById<MediaView>(R.id.media)
        // Set the Text.
        viewHolder.viewAdTag?.text = ad.sponsoredTranslation
        viewHolder.textViewTitle?.text = ad.advertiserName
        viewHolder.textViewSubtitle?.visibility = if (TextUtils.isEmpty(ad.adSocialContext)) View.GONE else View.VISIBLE
        viewHolder.textViewSubtitle?.text = ad.adSocialContext
        viewHolder.textViewBody?.visibility = if (TextUtils.isEmpty(ad.adBodyText)) View.GONE else View.VISIBLE
        viewHolder.textViewBody?.text = ad.adBodyText
        viewHolder.buttonAction?.visibility = if (ad.hasCallToAction()) View.VISIBLE else View.GONE
        viewHolder.buttonAction?.text = ad.adCallToAction

        // Add the AdOptionsView
        val adOptionsView = AdOptionsView(context, ad, adView.findViewById(R.id.native_ad_layout))
        viewHolder.adOptionsPlaceHolder?.removeAllViews()
        viewHolder.adOptionsPlaceHolder?.addView(adOptionsView)

        // Register the Title and CTA button to listen for clicks.
        val clickableViews = ArrayList<View>()
        clickableViews.add(adContainer)
        viewHolder.buttonAction?.let {
            clickableViews.add(it)
        }
        if (ad is NativeAd) {
            ad.registerViewForInteraction(adContainer, nativeAdMedia, nativeAdIcon, clickableViews)
        } else if (ad is NativeBannerAd) {
            ad.registerViewForInteraction(adContainer, nativeAdIcon, clickableViews)
        }

        if (adView.parent != null && adView.parent is ViewGroup) {
            (adView.parent as ViewGroup).removeView(adView)
        }
        adContainer.addView(adView)
    }
}